BCIF new format support

ValentinGuillaume opened the pull request:
The goal of this PR is to add our new bcif read/write feature to the main chemfiles repository.

Bcif file format is the new standard for downloading protein structure from the rcsb pdb since july 2024 and we would like to support it. @sguionni wrote an issue a few weeks ago regarding this specific subject 

The implementation should follow the pattern that is being used in the repository and doesn't need any new dependencies. It is tested by the read, write, reread method on a few heterogeneous files found on the rcsb pdb. The files are subject to their own PR on the associated repository.

ValentinGuillaume committed:
Implementing bcif format

ValentinGuillaume committed:
Reversing cmakelists target repo for file downloading

Luthaf commented:
Thanks a lot for this PR! I'm bit underwater right now, but I should be able to review by the end of next week.

In the mean time I can restart CI multiple time, until is passes!

ValentinGuillaume committed:
Updating main file repository commit sha and file hash

ValentinGuillaume commented:
No worries, we shall take our time with this ;)
In the meantime, I needed to have the test-data repo merged to be fully ready for this PR. Now that you approved it, I updated the hashes from the tests/CMakelists.txt to point toward the right commit and now it should work.

ValentinGuillaume committed:
Fixing test-data sha1sum value

ValentinGuillaume committed:
Restore CI compatible test-data file path

ValentinGuillaume committed:
Re-phrasing bcif error messages

ValentinGuillaume commented:
I resolved some issues regarding the checksum, file copy path and error message format enforcement.
Can you restart the CI ?

ValentinGuillaume committed:
Adding support for struct_conn reading/writing

ValentinGuillaume committed:
Removing ui64 literal integer declaration

ValentinGuillaume commented:
I added support for a table that connect some entities together sometimes. Also I removed the ui64 notation that caused you CI to fail compilation. 
Can you restart the CI ?

ValentinGuillaume committed:
implementing inter nucleic acid bonding

ValentinGuillaume commented:
The coverage issue is a big blow for me. The lines that need to be hit for the 83% to be reached are mostly failsafe lines that should occurs when reading a ill-formed or extremely huge bcif file. I have not such file at hand and thus increasing the coverage demands from me to craft those files, which is not as easy as crafting a cif or pdb or whatever human readable format. 
I will need way more time to do it right so for the time being, I will put this PR on hold. 
Do you have a doc/tutorial somewhere on how I can run this coverage test by myself before submitting this PR back ?

Luthaf commented:
Do not worry too much about the coverage for now, I can also lower the threshold. If you have some time it would be nice to if there are parts of your code are not tested and also not failsafe lines, and add tests for these. Error cases for invalid files are fine to leave as is for now.

ValentinGuillaume committed:
Improving bcif frame reading performances

ValentinGuillaume committed:
Cleaning up deprecated code

ValentinGuillaume committed:
Multi-threading buffer decoding

ValentinGuillaume committed:
WIP - starting string vector optimization

ValentinGuillaume committed:
Implementing better string management from bcif

ValentinGuillaume committed:
Restoring single thread reading with string vector optimization

ValentinGuillaume committed:
Removing unused code

ValentinGuillaume committed:
Cleaning up commented code. Improving test coverage

ValentinGuillaume commented:
I made some read speed improvements and tried to improve test coverage. Can you run the CI again ?

Luthaf commented:
The test faillure makes it look like the code is not pulling the most recent version of the test data

ValentinGuillaume committed:
Updating test-files hashes

ValentinGuillaume commented:
I updated the hashes to point toward the last commit that should contains all the needed files.
Can you restart the CI ?

ValentinGuillaume committed:
Enforcing types hoping it solves 32bits shenanigans

ValentinGuillaume committed:
MAXSIZE_T to SIZE_MAX

ValentinGuillaume committed:
codcov 3985

ValentinGuillaume committed:
codcov 3594

ValentinGuillaume committed:
Moving some code in BCIF_impl to improve code coverage

ValentinGuillaume committed:
Moving more private free functions to BCIF_impl for individual testing

ValentinGuillaume commented:
I made some change in the architecture in order to increase the code coverage value. Lets see if that's enough to pass that 83% limit

ValentinGuillaume committed:
8 more codecov lines

ValentinGuillaume commented:
Ok I added 8 easy lines of code coverage and we should be ready to merge.

Luthaf commented:
This is quite a lot of code, so I'll do a review ASAP, hopefully by next week. 

One question that jumps to mind after a first very high level look at the code: why do you have separate `BCIF` and `BCIF_impl` files? Just to be able to test the internal implementation?

Also, if you set any atom/residue/frame properties, please document them here: https://github.com/chemfiles/chemfiles/tree/master/doc/src/properties

ValentinGuillaume commented:
> Just to be able to test the internal implementation?

Well, yes, partly. I needed to unit test them to avoid forging binary data files and filling the chemfiles-test-data repo. And it was functions that are designed to work with the BCIF reading/writing code specifically, so I didn't want to expose them in the BCIF header because it would send the wrong message. Furthermore, I figured that BCIF.cpp was quite large too. 
All of that made me think it was cleaner to add this file. 
But feel free to move it to another location, rename it, or whatever suits your taste :) 

Regarding the tolm files and the properties : I followed the mmCIF guidelines as the bcif is designed as a binary counterpart of mmcif and thus have a lot in common with it. Did I miss something ?

Luthaf commented:
> Regarding the tolm files and the properties : I followed the mmCIF guidelines as the bcif is designed as a binary counterpart of mmcif and thus have a lot in common with it. Did I miss something ?

These files are documenting what properties are set/used when reading/writing different format. I guess you can just add a copy of the `mmCIF` line with `BCIF` instead! I'll let you check that this includes everything relevant

ValentinGuillaume committed:
Adding BCIF residue properties entries

ValentinGuillaume committed:
Adding BCIF atom properties entry

ValentinGuillaume committed:
The last residue property entry

ValentinGuillaume committed:
Repairing bcif related doc

ValentinGuillaume committed:
Improving doc script for windows compatibility

ValentinGuillaume commented:
I couldn't get the doc to work locally as I'm working on windows and the default text-file `open()` method from python uses os encoding, which made the angtrom symbol fail. 
Since the encoding will always be utf-8 I figured it was an appreciable fix, plus I could test it locally thanks to this modification.
Now the doc-build should work.

ValentinGuillaume committed:
Removing memory copy when parsing bcif from memory

Luthaf commented on src/formats/BCIF_impl.cpp:
```diff
@@ -0,0 +1,318 @@
+    else if (type_code == 8) type_str = "Uint32";
+    else if (type_code == 32) type_str = "Float32";
+    else if (type_code == 33) type_str = "Float64";
+    else type_str = "Int32";  // fallback
```
Why does this need a fallback? Would it be better to throw an error on unknown type code?

ValentinGuillaume replied:
I'll have an educated answer that once I encounter this case-scenario, but for now either way is fine for me. If you want I can make it throw instead.

Luthaf replied:
Yes I'd rather be safe and not try to guess here =)

Luthaf commented on src/formats/BCIF.cpp:
```diff
@@ -0,0 +1,3767 @@
+        }
+    };
+}
+namespace chemfiles {
```
Why is this in the chemfiles namespace? If it is only a utility class it can be defined inline in this file.

ValentinGuillaume replied:
You right, maybe this is something cleanup-worthy that I overlooked. I'll change that.

Luthaf commented on src/formats/BCIF.cpp:
```diff
@@ -0,0 +1,3767 @@
+    };
+}
+namespace chemfiles {
+    class StringCollection
```
Could you add some documentation about what this class is and what it does?

Luthaf commented on src/formats/BCIF.cpp:
```diff
@@ -0,0 +1,3767 @@
+                    conn.ptnr2.label_seq_id = (i < struct_data.ptnr2.auth_seq_id.size()) ? struct_data.ptnr2.auth_seq_id[i] : -1;
+                    conn.ptnr2.label_atom_id = (i < struct_data.ptnr2.label_atom_id.size()) ? struct_data.ptnr2.label_atom_id[i] : "";
+                    conn.pdbx_value_order = (i < struct_data.pdbx_value_order.size()) ? struct_data.pdbx_value_order[i] : "";
+                    // WIP
```
Why WIP?

ValentinGuillaume replied:
At some point it was, but I didn't remove the comment, sorry for that.

Luthaf commented on src/formats/BCIF.cpp:
```diff
@@ -0,0 +1,3767 @@
+                out.reserve(num_conns);
+                for (size_t i = 0; i < num_conns; ++i) {
+                    out.push_back({});
+                    BCIFData::StructConn& conn = out.back();
```
This is not very easy to read/understand. Could you add some comments that explain what's going on?

Luthaf commented on src/formats/BCIF.cpp:
```diff
@@ -0,0 +1,3767 @@
+                uint64_t bits;
+                std::memcpy(&bits, &data[i], sizeof(double));
+
+                // Store as little-endian
```
does BCIF specify endianess in any way?

ValentinGuillaume replied:
"specify" is a big word, but from what I could understand from the reference implementation [here](https://github.com/molstar/BinaryCIF), bytes are always little-endian in bcif.

Luthaf commented on src/formats/BCIF.cpp:
```diff
@@ -0,0 +1,3767 @@
+
+        has_written_ = true;
+    }
+    void BCIFFormat::decode(const char* data // TODO: When upgrading to c++20, use a span instead.
```
This function decodes everything and stores the data inside the class, right?

ValentinGuillaume replied:
The decode method parse the input buffer and organize data into the BCIFData that is owned by the instance of BCIFFormat for later usage.

Luthaf reviewed (commented):
The code looks very good overall, I left a couple of questions below.

For most other formats in chemfiles, the code that is currently in BCIF_impl (and part of formats/BCIF.cpp) would be handled by a custom `BinaryFile` class, that deals with reading, unpacking/packing data, … Do you have some bandwidth to refactor this code in this way, splitting better the bits that are mostly IO related from the bits that parse an actual chemical structure?

ValentinGuillaume commented:
> For most other formats in chemfiles, the code that is currently in BCIF_impl (and part of formats/BCIF.cpp) would be handled by a custom BinaryFile class, that deals with reading, unpacking/packing data, … Do you have some bandwidth to refactor this code in this way, splitting better the bits that are mostly IO related from the bits that parse an actual chemical structure?

Half of BCIF_impl free functions are unrelated to binary file io operations. So in the context of your question, please keep in mind that the class you propose won't fulfill all free functions responsibilities and won't erase the BCIF_impl.hpp entirely. 

The custom BinaryFile class raises two concerns on my end : 

1. The functions used to parse the bytes are tightly coupled with how the bcif format organizes the data. Sometimes it is rather simple and intrinsically generic (e.g. integers) sometimes it is very specific (e.g. strings). So my concern is the following : will I end up making a generic class that only decode very specific data ?
2. Do you have an usage of this potential BinaryFile class in mind ? If you do, it would help me to understand the relevance of this refactoring and maybe the responsibilities this class would have. I have trouble to figure what the class would do and why we need it to be a class rather than a set of free functions.

ValentinGuillaume committed:
Commenting BCIF.cpp more and changing namespace of StringCollection

ValentinGuillaume committed:
Making get_int_type_str throw for unknown types.

ValentinGuillaume commented:
Following your advices, I commented more and made the get_int_type_str function throw for unknown type. 
(I also removed the include span that I forgot to cleanup and caused gcc9 to fail)

ValentinGuillaume committed:
Improving integer type support. Adding failing test on 4hhb.

ValentinGuillaume committed:
Fixing failing test

ValentinGuillaume committed:
Adding failing tests on 6fxo

ValentinGuillaume committed:
Solving residue gap bonding issue

ValentinGuillaume committed:
Updating bcif residue reading toward auth_res_id

ValentinGuillaume committed:
Updating number of expected residues

Luthaf commented:
> The custom BinaryFile class raises two concerns on my end :
> 
>     1. The functions used to parse the bytes are tightly coupled with how the bcif format organizes the data. Sometimes it is rather simple and intrinsically generic (e.g. integers) sometimes it is very specific (e.g. strings). So my concern is the following : will I end up making a generic class that only decode very specific data ?
> 
>     2. Do you have an usage of this potential BinaryFile class in mind ? If you do, it would help me to understand the relevance of this refactoring and maybe the responsibilities this class would have. I have trouble to figure what the class would do and why we need it to be a class rather than a set of free functions.

This would not be a generic class, but a BCIF-specific class. For example, we have a [NetCDF3](https://github.com/chemfiles/chemfiles/blob/master/src/files/Netcdf3File.cpp) binary file implementation, that provides tools for reading/writing files in NetCDF3 format; and is then used by the [AmberNetCDF](https://github.com/chemfiles/chemfiles/blob/master/src/formats/AmberNetCDF.cpp) format to implement this format. 

The file class is only used together with one format class, it's more about separating concerns between "extracting data from a stream of bytes" and "interpreting this data as a chemical structure". 

Is this clearer or not really?

ValentinGuillaume committed:
Fixing issue where a vector of 0's wasn't counted as invalid data for residue_id

ValentinGuillaume committed:
adding metadata reading

ValentinGuillaume committed:
Facto string reading

ValentinGuillaume committed:
Adding local test for label_alt_id bonding

ValentinGuillaume committed:
Adding support for altloc
